src/java-interop/java-interop-dlfcn.cc

```diff
@@ -0,0 +1,173 @@
+#include "java-interop.h"
+#include "java-interop-dlfcn.h"
+#include "java-interop-util.h"
+
+#ifdef WINDOWS
+#include <libloaderapi.h>
+#include <winerror.h>
+#include <wtypes.h>
+#include <winbase.h>
+#else
+#include <dlfcn.h>
+#endif
+
+static char *
+_get_last_dlerror ()
+{
+#ifdef WINDOWS
+
+	DWORD error = GetLastError ();
+	if (error == ERROR_SUCCESS /* 0 */) {
+		return nullptr;
+	}
+
+	wchar_t *buf = nullptr;
+
+	DWORD size = FormatMessageW (
+			/* dwFlags */       FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM |FORMAT_MESSAGE_IGNORE_INSERTS,
+			/* lpSource */      NULL,
+			/* dwMessageId */   error,
+			/* dwLanguageId */  MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+			/* lpBuffer */      (LPWSTR) &buf,
+			/* nSize */         0,
+			/* Arguments */     NULL
+	);
+	if (size == 0)
+		return nullptr;
+
+	char *message = utf16_to_utf8 (buf);
+	LocalFree (buf);
+
+	return message;
+
+#else   // ndef WINDOWS
+
+	return java_interop_strdup (dlerror ());
+
+#endif  // ndef WINDOWS
+}
+
+void*
+java_interop_load_library (const char *path, unsigned int flags, char **error)
+{
+	if (error) {
+		java_interop_free (*error);
+		*error = nullptr;
+	}
+	if (path == nullptr) {
+		if (error)
+			*error = java_interop_strdup ("path=nullptr is not supported");
+		return nullptr;
+	}
+	if (flags != 0) {
+		if (error)
+			*error = java_interop_strdup ("flags has unsupported value");
+		return nullptr;
+	}
+
+	void *handle    = nullptr;
+
+#ifdef WINDOWS
+
+	wchar_t *wpath   = utf8_to_utf16 (path);
+	if (path == nullptr) {
+		if (error)
+			*error = java_interop_strdup ("could not convert path to UTF-16");
+		return nullptr;
+	}
+	HMODULE module  = LoadLibraryExW (
+			/* lpLibFileName */ wpath,
+			/* hFile */         nullptr,
+			/* dwFlags */       LOAD_LIBRARY_SEARCH_APPLICATION_DIR | LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR | LOAD_LIBRARY_SEARCH_USER_DIRS
+	);
+	java_interop_free (wpath);
+
+	handle = reinterpret_cast<void*>(module);
+
+#else   // ndef WINDOWS
+
+	handle  = dlopen (path, RTLD_LAZY | RTLD_LOCAL);
+
+#endif  // ndef WINDOWS
+
+	if (handle == nullptr && error) {
+		*error = _get_last_dlerror ();
+	}
+
+	return handle;
+}
+
+void*
+java_interop_get_symbol_address (void *library, const char *symbol, char **error)
+{
+	if (error) {
+		java_interop_free (*error);
+		*error = nullptr;
+	}
+
+	if (library == nullptr) {
+		if (error)
+			*error = java_interop_strdup ("library=nullptr");
+		return nullptr;
+	}
+	if (symbol == nullptr) {
+		if (error)
+			*error = java_interop_strdup ("symbol=nullptr");
+		return nullptr;
+	}
+
+	void *address   = nullptr;
```

Comment:
Could this use `uint8_t*` instead? The allocated memory is a collection of 8-bit bytes and personally I think `void*` should be avoided whenever possible. In this case there could be just one cast on return.

Reply:
Not sure I like this idea at all, considering that it's typically going to be an invocable function pointer.  Having that as `uint8_t*` doesn't seem…sane.

Reply:
`void*` pointers just make me shiver... :)

Reply:
…and I think they're more honest. :-)

Reply:
...until somebody forgets to cast them to the proper type and does pointer arithmetic... :)